tests/phpunit/tests/post/getPages.php

```diff
@@ -325,6 +325,141 @@ public function test_get_pages_include_exclude() {
 		$this->assertSame( $inc, $exc_result );
 	}
 
+	/**
+	 * @ticket 12821
+	 */
+	public function test_get_pages_test_filter() {
+		register_post_type( 'wptests_pt', array( 'hierarchical' => true ) );
+
+		$posts             = self::factory()->post->create_many(
+			2,
+			array(
+				'post_type' => 'wptests_pt',
+			)
+		);
+		$query_args_values = array();
+
+		// Filter the query to return the wptests_pt post type.
+		add_filter(
+			'get_pages_query_args',
+			static function( $query_args ) use ( &$query_args_values ) {
+				$query_args['post_type'] = 'wptests_pt';
+				$query_args_values       = $query_args;
+				return $query_args;
+			}
+		);
+
+		$pages    = get_pages();
+		$page_ids = wp_list_pluck( $pages, 'ID' );
+		$this->assertSameSets( $posts, $page_ids, 'The return post ids should match the post type wptests_pt.' );
+
+		$default_args = array(
+			'orderby'                => array( 'post_title' => 'ASC' ),
+			'order'                  => 'ASC',
+			'post__not_in'           => array(),
+			'meta_key'               => '',
+			'meta_value'             => '',
+			'posts_per_page'         => -1,
+			'offset'                 => 0,
+			'post_type'              => 'wptests_pt',
+			'post_status'            => array( 'publish' ),
+			'update_post_term_cache' => false,
+			'update_post_meta_cache' => false,
+			'ignore_sticky_posts'    => true,
+			'no_found_rows'          => true,
+		);
+
+		$this->assertSameSets( $default_args, $query_args_values, 'Query arguments should match expected values' );
+	}
+
+	/**
+	 * @ticket 12821
+	 * @covers ::get_pages
+	 * @dataProvider data_get_pages_args
+	 */
+	public function test_get_pages_args_test_filter( $args, $expected_query_args ) {
+		$filter = new MockAction();
+		add_filter( 'get_pages_query_args', array( $filter, 'filter' ), 10, 2 );
+
+		get_pages( $args );
+
+		$filter_args = $filter->get_args();
+
+		$default_args = array(
+			'orderby'                => array( 'post_title' => 'ASC' ),
+			'order'                  => 'ASC',
+			'post__not_in'           => array(),
+			'meta_key'               => '',
+			'meta_value'             => '',
+			'posts_per_page'         => -1,
+			'offset'                 => 0,
+			'post_type'              => 'page',
+			'post_status'            => array( 'publish' ),
+			'update_post_term_cache' => false,
+			'update_post_meta_cache' => false,
+			'ignore_sticky_posts'    => true,
+			'no_found_rows'          => true,
+		);
+
+		$parsed_args = wp_parse_args( $expected_query_args, $default_args );
+
+		$this->assertSameSets( $filter_args[0][0], $parsed_args );
```

Comment:
I believe this could be minimal test coverage for the second argument to the filter:

```suggestion
		$this->assertCount( 2, $filter_args[0] );
		$this->assertSameSets( $filter_args[0][0], $parsed_args );
		$this->assertSameSets( $filter_args[0][1], $parsed_args );
```

Reply:
Just updated the suggestion to swap the first/second args to `assertSameSets` since the expected arg is supposed to come first, and the actual arg second. Also added missing messages.